interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java

```diff
@@ -1736,57 +1738,100 @@ public void performSoakTest(
-            .withInterceptors(recordClientCallInterceptor(clientCallCapture));
-      }
-      SoakIterationResult result = 
-          performOneSoakIteration(soakStub, soakRequestSize, soakResponseSize);
-      SocketAddress peer = clientCallCapture
-          .get().getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
-      StringBuilder logStr = new StringBuilder(
-          String.format(
-              Locale.US,
-              "soak iteration: %d elapsed_ms: %d peer: %s server_uri: %s",
-              i, result.getLatencyMs(), peer != null ? peer.toString() : "null", serverUri));
-      if (!result.getStatus().equals(Status.OK)) {
-        totalFailures++;
-        logStr.append(String.format(" failed: %s", result.getStatus()));
-      } else if (result.getLatencyMs() > maxAcceptablePerIterationLatencyMs) {
-        totalFailures++;
-        logStr.append(
-            " exceeds max acceptable latency: " + maxAcceptablePerIterationLatencyMs);
-      } else {
-        logStr.append(" succeeded");
-      }
-      System.err.println(logStr.toString());
-      iterationsDone++;
-      latencies.recordValue(result.getLatencyMs());
-      long remainingNs = earliestNextStartNs - System.nanoTime();
-      if (remainingNs > 0) {
-        TimeUnit.NANOSECONDS.sleep(remainingNs);
-      }
+    Thread[] threads = new Thread[numThreads];
+    int soakIterationsPerThread = (int) Math.ceil((double) soakIterations / numThreads);
+    for (int threadInd = 0; threadInd < numThreads; threadInd++) {
+      final int startIteration = threadInd * soakIterationsPerThread;
+      final int endIteration = Math.min(startIteration + soakIterationsPerThread, soakIterations);
+      threads[threadInd] = new Thread(() -> {
+        ManagedChannel currentChannel = soakChannel;
+        TestServiceGrpc.TestServiceBlockingStub currentStub = soakStub;
+        for (int i = startIteration; i < endIteration; i++) {
+          if (System.nanoTime() - startNs >= TimeUnit.SECONDS.toNanos(overallTimeoutSeconds)) {
+            break;
+          }
+          long earliestNextStartNs = System.nanoTime()
+              + TimeUnit.MILLISECONDS.toNanos(minTimeMsBetweenRpcs);
+          if (resetChannelPerIteration) {
+            currentChannel.shutdownNow();
+            try {
+              currentChannel.awaitTermination(10, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+            }
+            currentChannel = createChannel();
+            currentStub = TestServiceGrpc
+                .newBlockingStub(soakChannel)
+                .withInterceptors(recordClientCallInterceptor(clientCallCapture));
+          }
+          SoakIterationResult result;
+          try {
+            result = performOneSoakIteration(currentStub, soakRequestSize, soakResponseSize);
+          } catch (Exception e) {
+            synchronized (this) {
+              totalFailures.incrementAndGet();
```

Comment:
we shouldn't need to examine `totalFailures` until the thread is done and joined.

We can have a list of per-thread counters that each thread tallies up while running. Then after joining all threads, the main thread can examine all of those counters and create a summary.